Import Override Support

pmyjavec opened the pull request:
Allow users to override the DML parameters found in dump/backup files.

This is especially useful when importing large databases as it doesn't
require the file to be extracted to change these values.

* Allow users to specify the destination database name.
* Allow users to override the retention policy value.

###### Required for all non-trivial PRs
- [ ] Rebased/mergable
- [ ] Tests pass
- [ ] CHANGELOG.md updated
- [ ] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

###### Required only if applicable
_You can erase any checkboxes below this note if they are not applicable to your Pull Request._
- [ ] [InfluxQL Spec](https://github.com/influxdata/influxdb/blob/master/influxql/README.md) updated
- [ ] Provide example syntax
- [ ] Update man page when modifying a command
- [ ] Config changes: update sample config (`etc/config.sample.toml`), server `NewDemoConfig` method, and `Diagnostics` methods reporting config settings, if necessary
- [ ] [InfluxData Documentation](https://github.com/influxdata/docs.influxdata.com): issue filed or pull request submitted \<link to issue or pull request\>

e-dard commented on cmd/influx/main.go:
```diff
@@ -56,6 +56,8 @@ func main() {
 	fs.IntVar(&c.ImporterConfig.PPS, "pps", defaultPPS, "How many points per second the import will allow.  By default it is zero and will not throttle importing.")
 	fs.StringVar(&c.ImporterConfig.Path, "path", "", "path to the file to import")
 	fs.BoolVar(&c.ImporterConfig.Compressed, "compressed", false, "set to true if the import file is compressed")
+	fs.StringVar(&c.ImporterConfig.RetentionPolicy, "retention-policy", "", "Overrides the retention policy when importing a databse from file")
```
`importing a databse`

e-dard replied:
I think that this, and the `destination-database` flag should say something like "writes the import data to the provided retention policy when using -import". `Overrides` seemed a little confusion to me as to whether it was modifying the import file or changing the location the imported data was written to.

e-dard commented on cmd/influx/main.go:
```diff
@@ -56,6 +56,8 @@ func main() {
 	fs.StringVar(&c.ImporterConfig.Path, "path", "", "path to the file to import")
 	fs.BoolVar(&c.ImporterConfig.Compressed, "compressed", false, "set to true if the import file is compressed")
+	fs.StringVar(&c.ImporterConfig.RetentionPolicy, "retention-policy", "", "Overrides the retention policy when importing a databse from file")
+	fs.StringVar(&c.ImporterConfig.DestinationDatabase, "destination-database", "", "Overrides the database name when importing a databse from file")
```
`importing a databse`

e-dard commented on cmd/influx/main.go:
```diff
@@ -96,6 +98,10 @@ func main() {
   -compressed
        Set to true if the import file is compressed
+  -retention-policy
+		Overrides the retention policy when importing a databse from file. Must be used with -import.
```
`importing a databse`

e-dard commented on cmd/influx/main.go:
```diff
@@ -96,6 +98,10 @@ func main() {
        Set to true if the import file is compressed
+  -retention-policy
+		Overrides the retention policy when importing a databse from file. Must be used with -import.
+  -destination-database
```
`-destination-databse`

e-dard commented on importer/v8/importer.go:
```diff
@@ -153,18 +157,37 @@ func (i *Importer) processDDL(scanner *bufio.Scanner) {
 
-func (i *Importer) processDML(scanner *bufio.Scanner) {
+// processDML actually processes each of the points once it has created the target database. The database
+// databae can either be specified by the user, or it will be read from the import file.
```
`databae` can go.

e-dard commented on importer/v8/importer.go:
```diff
@@ -153,18 +157,37 @@ func (i *Importer) processDDL(scanner *bufio.Scanner) {
-func (i *Importer) processDML(scanner *bufio.Scanner) {
+// processDML actually processes each of the points once it has created the target database. The database
+// databae can either be specified by the user, or it will be read from the import file.
+//The saame goes for the retention policy it too can be overriden by specifying `rp` or else it too is set
```
`saame`

e-dard commented on man/influx.txt:
```diff
@@ -78,4 +78,10 @@ OPTIONS
+  Overrides the database name found in the import file and sets it to the specified value, must be used with -import.
+
+-retention-policy::
+  Overrides the retention policy when importing a databse from file. must be used with -import.
```
`databse`

pmyjavec commented:
Ok, 

@e-dard all issues have been fixed.

e-dard commented on importer/v8/importer.go:
```diff
@@ -153,18 +157,36 @@ func (i *Importer) processDDL(scanner *bufio.Scanner) {
 }
 
-func (i *Importer) processDML(scanner *bufio.Scanner) {
+// processDML actually processes each of the points once it has created the target database. The databae can
```
`databae`

e-dard commented:
@pmyjavec this PR will also need one more review from a platform team member.

pmyjavec commented:
@e-dard,

Ok no problem, thanks for that. Sorry about the typos, I'll blame it on my new ergonomic keyboard :)

corylanou commented on cmd/influx/main.go:
```diff
@@ -56,6 +56,8 @@ func main() {
 	fs.IntVar(&c.ImporterConfig.PPS, "pps", defaultPPS, "How many points per second the import will allow.  By default it is zero and will not throttle importing.")
 	fs.StringVar(&c.ImporterConfig.Path, "path", "", "path to the file to import")
 	fs.BoolVar(&c.ImporterConfig.Compressed, "compressed", false, "set to true if the import file is compressed")
+	fs.StringVar(&c.ImporterConfig.RetentionPolicy, "retention-policy", "", "Writes the import data to the specified retention policy when using -import.")
```
for consistency, should it be `destination-retention-policy` to match `destination-database`?

pmyjavec replied:
Hello @corylanou,

I did think about using that parameter name, I'm pretty big consistency too but I think it's really a little too long and considering it's not used for anything else, nor I can't think of any case where it ever will, I felt that this name is probably fine.

Thoughts?

pmyjavec replied:
@corylanou, rather than updating the README, how about we just change the wording to something like:

> Overrides all retention policy statements found in the import file and sets them to the specified parameter when using -import

Something similar for the `-destination-database` parameter.

pmyjavec replied:
@e-dard, @corylanou any thoughts?

corylanou commented on cmd/influx/main.go:
```diff
@@ -96,6 +98,10 @@ func main() {
        Path to file to import
   -compressed
        Set to true if the import file is compressed
+  -retention-policy
```
we we do update the parameter per my comment above, this reference would need to be updated as well

corylanou commented on man/influx.txt:
```diff
@@ -78,4 +78,10 @@ OPTIONS
 -version::
   Outputs the version of the influx client.
 
+-retention-policy::
```
may need to update this reference if we update the name.

stromnet commented:
Any ETA on this? Just updated to 1.3 in the hope to see this landed but unfortunately not.. :)

rbetts commented on importer/v8/importer.go:
```diff
@@ -153,18 +157,36 @@ func (i *Importer) processDDL(scanner *bufio.Scanner) {
 		}
-		i.queryExecutor(line)
+
+		i.createDatabaseQuery = line
```
This constantly sets `createDatabaseQuery` to the last line of the DDL section. Is that correct? Does `CREATE DATABASE` have to be the final line of the DDL preamble?

Also, whatever `line` is -- it was executed before being stored here; meaning at minimum, we will CREATE the database in the DDL section and then create the database for the override; that seems incorrect. If an override is specified, I wouldn't expect the database in the file to be created.

rbetts commented on importer/v8/importer.go:
```diff
@@ -153,18 +157,36 @@ func (i *Importer) processDDL(scanner *bufio.Scanner) {
+		i.createDatabaseQuery = fmt.Sprintf("CREATE DATABASE %s", dboverride)
+		i.database = dboverride
+	}
+	i.queryExecutor(i.createDatabaseQuery)
```
If `dboverride == ""`, this command was already executed in `processDDL()`.  Also, I don't expect `CREATE DATABASE` to be called in a function named `processDML()`.

pmyjavec replied:
I'm unclear on how that happens because the processDDL just finds the create database statement and sets it for processing later. It used to be the case that processDDL creates the database but in the PR that is no longer the case. From the previous snippet:

` i.createDatabaseQuery = line `

pmyjavec replied:
Notice that it just sets `i.createDatabaseQuery`, it does not execute the queryExecutor.

pmyjavec commented:
Hey @rbetts,

I've addressed your concerns by removing any database creation statements when the overrides are specified. I think this makes more sense anyway.

Thanks
